### PR TITLE
Arr integration

### PR DESCRIPTION
**Pull request by tachukao**
1. allow users to integrate Ndarrays of dimension n. the results is an Ndarray of dimension (n+1), where the first dimension corresponds to time
2. expose type `tspec` and `Solver` directly in `Owl_ode`

### CODE REVIEW

**Comment by tachukao** [comment]
argh...Travis is failing because the current version of `owl` on `opam` doesn't have `expand` in `owl_types_ndarray_basic.ml` yet

**Comment by ryanrhymes** [comment]
@mseri we probably need a new release of Owl?

**Comment by mseri** [comment]
I can draft one later today

**Comment by ryanrhymes** [comment]
Awesome!

**Comment by mseri** [comment]
There are some inconsistencies with the tests. I’ll probably make a release with testing disabled for now

**Comment by mseri** [comment]
In progress: https://github.com/ocaml/opam-repository/pull/15278

File: src/ode/owl_ode.ml [hunk 1]
```diff
@@ -3,3 +3,4 @@ module Ode = Owl_ode_base.Ode
 module Types = Owl_ode_base.Types
 module Native = Native
 module Symplectic = Symplectic
+include Types
```
Comment (mseri) [inline src/ode/owl_ode.ml]: How much will this mess up the types? Types from different modules would become different right? I'd rather keep it spearate as before

File: src/sundials/owl_ode_sundials.ml [hunk 2]
```diff
@@ -76,26 +70,25 @@ let cvode_s'
...
+  let session, yvec, y = make_cvode_session ~stiff ~relative_tol ~abs_tol f y0 t0 in
   Cvode.set_stop_time session tstop;
   fun _y t ->
     let tout = t +. dt in
     let rec until tout =
       let t', r = Cvode.solve_normal session tout yvec in
       match r with
       | Success ->
-        let y' = Mat.copy (unwrap (dim1, dim2) y) in
+        let y' = Arr.copy (Arr.reshape (unwrap y) s) in
```
Comment (mseri) [inline src/sundials/owl_ode_sundials.ml:90]: Can you explain me more about the rationale? Why is mixing up the types in this way, sometimes mat and sometimes arr, better?

File: src/ode/owl_ode.ml (hunk 1 above)
Comment (tachukao) [reply src/ode/owl_ode.ml]: yea that's true I didn't think of that. I'll revert it

File: src/sundials/owl_ode_sundials.ml (hunk 2 above)
Comment (tachukao) [reply src/sundials/owl_ode_sundials.ml:90]: the rationale here is to support integration of higher dimensional Ndarrays (n >= 2).  `Mat.mat` and `Arr.arr` are both `genarray` underneath.

File: src/ode/owl_ode.ml (hunk 1 above)
Comment (mseri) [reply src/ode/owl_ode.ml]: If you want, you could add `owl_ode.mli` and expose the types there by hand. In the implementation you would just have `val tspec = Types.tspec` while in the signature you can fully expand it (you will also have to copy-paste the docstring but I guess we can live with some level of duplication). I think that should work.

**Comment by mseri** [comment]
We should also update the lower bounds on owl and owl-base in the opam files, and the examples

File: src/ode/owl_ode.ml (hunk 1 above)
Comment (tachukao) [reply src/ode/owl_ode.ml]: argh of course...that would be the way to do it. But perhaps, this is best left for another PR.

Comment (mseri) [reply src/ode/owl_ode.ml]: I don’t know. It’s such a small change that we may as well add it here. Do you think it would take too long?

Comment (tachukao) [reply src/ode/owl_ode.ml]: wouldn't we have to change the docs as well??

Comment (tachukao) [reply src/ode/owl_ode.ml]: but yea to be fair, we'll probably have to update the docs anyway. okay, I'll make the changes expose the type `tspec`

Comment (mseri) [reply src/ode/owl_ode.ml]: I have added a commit that solves this issue. I have also rebased on the current origin to cleanup the history from the many merges

**Pull request merged**